test/IntegrationImport.hs

```diff
@@ -0,0 +1,53 @@
+module IntegrationImport
+    ( createDBPool
+    , clickOn
+    , fillIn
+    , inLoginWindow
+    , waitForElem
+    , waitForText
+    , module X
+    ) where
+
+import ClassyPrelude as X hiding (Element)
+import Test.Hspec.WebDriver as X hiding (shouldBe, shouldReturn)
+
+import Settings
+import TestImport (getAppSettings, wipeDBPool)
+import Test.WebDriver.Class
+import Test.WebDriver.Commands.Wait
+
+import Control.Monad.Logger
+import Database.Persist.Postgresql
+
+inLoginWindow :: WebDriver m => m a -> m ()
+inLoginWindow f = do
+    cw <- getCurrentWindow
+    focusWindow =<< return . unsafeHead . filter (/= cw) =<< windows
+    void $ f
+    focusWindow cw
+
+clickOn :: WebDriver m => Selector -> m ()
+clickOn = click <=< waitForElem
+
+fillIn :: WebDriver m => Selector -> Text -> m Element
+fillIn x value = do
+    input <- findElem x
+    sendKeys value input
+    return input
+
+waitForText :: WebDriver m => Selector -> m Text
+waitForText = getText <=< waitForElem
+
+waitForElem :: WebDriver m => Selector -> m Element
+waitForElem = waitUntil 50000 . findElem
```

Comment:
I'm not sure, I think it's whatever threadDelay takes.

I think excessive is good, it'll only be slow in the failure case. I'd
rather that then run the risk of not waiting long enough and getting a
false failure.
On May 1, 2015 11:35 AM, "Joe Ferris" <notifications@github.com> wrote:

> In test/IntegrationImport.hs
> <https://github.com/thoughtbot/carnival/pull/250#discussion_r29508659>:
>
> > +    focusWindow cw
> > +
> > +clickOn :: WebDriver m => Selector -> m ()
> > +clickOn = click <=< waitForElem
> > +
> > +fillIn :: WebDriver m => Selector -> Text -> m Element
> > +fillIn x value = do
> > +    input <- findElem x
> > +    sendKeys value input
> > +    return input
> > +
> > +waitForText :: WebDriver m => Selector -> m Text
> > +waitForText = getText <=< waitForElem
> > +
> > +waitForElem :: WebDriver m => Selector -> m Element
> > +waitForElem = waitUntil 50000 . findElem
>
> What is the 50000 value here? Milliseconds? Is this excessive?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thoughtbot/carnival/pull/250/files#r29508659>.
>